src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java

```diff
@@ -236,24 +275,39 @@ Object sanitizeListAndRecordMutation(@Nonnull List objects, @CheckForNull EnvVar
     Object sanitizeObjectAndRecordMutation(@CheckForNull Object o, @CheckForNull EnvVars vars) {
         // Package scoped so we can test it directly
         Object tempVal = o;
+        DescribableModel m = null;
         if (tempVal instanceof Step) {
             // Ugly but functional used for legacy syntaxes with metasteps
+            m = DescribableModel.of(tempVal.getClass());
             tempVal = ((Step)tempVal).getDescriptor().defineArguments((Step)tempVal);
         } else if (tempVal instanceof UninstantiatedDescribable) {
             tempVal = ((UninstantiatedDescribable)tempVal).toMap();
+        } else if (tempVal instanceof Describable) {  // Raw Describables may not be safe to store, so we should explode it
+            m = DescribableModel.of(tempVal.getClass());
+            tempVal = m.uninstantiate2(o).toMap();
         }
 
         if (isOversized(tempVal)) {
             this.isUnmodifiedBySanitization = false;
             return NotStoredReason.OVERSIZE_VALUE;
         }
 
+        if (!isStoreableType(tempVal)) {  // If we're not a legal type to store, then don't.
+            this.isUnmodifiedBySanitization = false;
+            return NotStoredReason.UNSERIALIZABLE;
+        }
+
         Object modded = tempVal;
         if (modded instanceof Map) {
             // Recursive sanitization, oh my!
             modded = sanitizeMapAndRecordMutation((Map)modded, vars);
         } else if (modded instanceof List) {
             modded = sanitizeListAndRecordMutation((List) modded, vars);
+        } else if (modded != null && modded.getClass().isArray()) {
```

Comment:
Not sure why we are dealing with arrays here—`structs` deals with `List`s, or at least normalizes to them, but perhaps you are still working with unprocessed direct input from `StepDescriptor.instantiate`?

Reply:
Yes, my assumption is that it *might* be possible -- or in any case, that it's probably safer to deal with it than not to.